lib/cocoa/src/protocol/TCompactProtocol.m

```diff
@@ -0,0 +1,706 @@
+    }
+}
+
+-(void)writeListBeginWithElementType:(int)elementType size:(int)size
+{
+    [self writeCollectionBeginWithElementType:elementType size:size];
+}
+
+-(void)writeSetBeginWithElementType:(int)elementType size:(int)size
+{
+    [self writeCollectionBeginWithElementType:elementType size:size];
+}
+
+-(void)writeBool:(BOOL)b
+{
+    if (boolFieldId != nil && boolFieldName != nil && boolFieldType != nil)
+    {
+        // we haven't written the field header yet
+        [self writeFieldBeginInternalWithName:boolFieldName type:[boolFieldType intValue] fieldID:[boolFieldId intValue] typeOverride:b ? TCType_BOOLEAN_TRUE : TCType_BOOLEAN_FALSE];
+        
+        [boolFieldId release_stub];
+        [boolFieldName release_stub];
+        [boolFieldType release_stub];
+        
+        boolFieldId = nil;
+        boolFieldName = nil;
+        boolFieldType = nil;
+    }
+    else
+    {
+        // we're not part of a field, so just Write the value.
+        [self writeByteDirect:b ? TCType_BOOLEAN_TRUE : TCType_BOOLEAN_FALSE];
+    }
+}
+
+-(void)writeByte:(uint8_t)value
+{
+    [self writeByteDirect:value];
+}
+
+-(void)writeI16:(short)value
+{
+    [self writeVarint32:[self i32ToZigZag:value]];
+}
+
+-(void)writeI32:(int32_t)value
+{
+    [self writeVarint32:[self i32ToZigZag:value]];
+}
+
+-(void)writeI64:(int64_t)value
+{
+    [self writeVarint64:[self i64ToZigZag:value]];
+}
+
+-(void)writeDouble:(double)value
+{
+    //Safe bit-casting double->uint64
+    uint8_t data[8];
+    uint64_t bits = 0;
+    memcpy(data, &value, 8);
+    memcpy(&bits, data, 8);
+    
+    bits = OSSwapHostToLittleInt64(bits);
+    [mTransport write:(uint8_t*)&bits offset:0 length:8];
+}
+
+-(void)writeString:(NSString*)value
+{
+    [self writeBinary:[value dataUsingEncoding:NSUTF8StringEncoding]];
+}
+
+-(void)writeBinary:(NSData*)data
+{
+    [self writeVarint32:(uint32_t)data.length];
+    [mTransport write:data.bytes offset:0 length:data.length];
+}
+
+-(void)writeMessageEnd {}
+-(void)writeMapEnd {}
+-(void)writeListEnd {}
+-(void)writeSetEnd {}
+-(void)writeFieldEnd {}
+
+-(void)writeCollectionBeginWithElementType:(int)elementType size:(int)size
+{
+    if (size <= 14)
+    {
+        [self writeByteDirect:size << 4 | [self getCompactTypeForTType:elementType]];
+    }
+    else
+    {
+        [self writeByteDirect:0xf0 | [self getCompactTypeForTType:elementType]];
+        [self writeVarint32:(uint32_t)size];
+    }
+}
+
+-(void)writeVarint64:(UInt64)n
+{
+    uint8_t varint64out[10] = {0};
+    
+    int idx = 0;
+    while (true)
+    {
+        if ((n & ~0x7FL) == 0)
+        {
+            varint64out[idx++] = (uint8_t)n;
+            break;
+        }
+        else
+        {
+            varint64out[idx++] = (uint8_t)((n & 0x7F) | 0x80);
+            n >>= 7;
+        }
+    }
+    [mTransport write:varint64out offset:0 length:idx];
+}
+
+-(uint32_t)i32ToZigZag:(int32_t)n
```

Comment:
It would be great to have a doc comment with a link to a page documenting what a zig-zag is. I noticed that the node library does explain in a comment what it is and why it's used; without some documentation, this is rather mysterious, especially if you look only at the Obj-C code.